### PR TITLE
Accomodate BE platforms. Make sure we always write flatbuffers LE

### PR DESCRIPTION
**Pull request by yuslepukhin**
### Description

flatbuffers always write data in LE and it is automatically traslated to/from BE as needed,
but only if we use proper accessors. This would work for shape.
However, we store parameters as bytes, so we need to swap bytes as needed for BE.

### Motivation and Context
Address https://github.com/microsoft/onnxruntime/issues/22364

### CODE REVIEW

File: onnxruntime/lora/adapter_format_utils.cc [hunk 1]
```diff
@@ -75,35 +77,82 @@ const Adapter* ValidateAndGetAdapterFromBytes(gsl::span<const uint8_t> bytes) {
...
+  } else {
+    data_vec = flat_builder.CreateVector(data.data(), data.size());
+  }
   fbs_tensor = CreateParameter(flat_builder, name_str, shape_vec, data_type, data_vec);
 }
 
+template <class T>
+struct ReadDataForBigEndian {
+  Status operator()(gsl::span<const uint8_t> src, Tensor& dst) const {
+    auto src_span = ReinterpretAsSpan<const unsigned char>(src);
```
Comment (skottmckay) [inline onnxruntime/lora/adapter_format_utils.cc:117]: nit: probably doesn't matter but could we do the data type conversion from uint8_t to unsigned char outside of the templatized code?

File: onnxruntime/lora/adapter_format_utils.cc [hunk 2]
```diff
@@ -75,35 +77,82 @@ const Adapter* ValidateAndGetAdapterFromBytes(gsl::span<const uint8_t> bytes) {
...
+
+    auto status = disp.InvokeRet<Status, ReadDataForBigEndian>(src_span, tensor);
+    ORT_THROW_IF_ERROR(status);
+    Tensor::InitOrtValue(std::move(tensor), result);
+  } else {
+    // const_cast is necessary due to Tensor class API
+    Tensor::InitOrtValue(elem_type,
+                         TensorShape(shape),
+                         const_cast<uint8_t*>(param.raw_data()->data()),
+                         cpu_allocator->Info(),
```
Comment (skottmckay) [inline onnxruntime/lora/adapter_format_utils.cc:153]: Can we avoid creating the local cpu_allocator on the LE path and use cpu_meminfo here?

File: onnxruntime/lora/adapter_format_utils.cc
Comment (skottmckay) [inline onnxruntime/lora/adapter_format_utils.cc:1]: Can someone with a BE machine validate the changes?

**Review by skottmckay — APPROVED** [review]

File: onnxruntime/lora/adapter_format_utils.cc [hunk 3]
```diff
@@ -75,35 +77,108 @@ const Adapter* ValidateAndGetAdapterFromBytes(gsl::span<const uint8_t> bytes) {
...
 
-  auto elem_type = DataTypeImpl::TensorTypeFromONNXEnum(static_cast<int32_t>(data_type))->GetElementType();
-  // const_cast is necessery due to Tensor class API
-  Tensor::InitOrtValue(elem_type,
-                       TensorShape(shape_span),
-                       const_cast<uint8_t*>(param.raw_data()->data()),
-                       cpu_meminfo,
-                       result);
+  if constexpr (endian::native == endian::big) {
+    if (elem_type->Size() > 1) {
```
Comment (tianleiwu) [inline onnxruntime/lora/adapter_format_utils.cc:162]: I am curious about the handling of 4 bit data types like UINT4, INT4 and FLOAT4E2M1.

**Review by tianleiwu — APPROVED** [review]

**Comment by ranjitshs** [comment]
@yuslepukhin 
Verified the latest changes. It's looks fine.

**Pull request merged**